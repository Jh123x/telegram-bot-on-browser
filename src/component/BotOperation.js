import { useState, useEffect } from "react";
import { Fragment } from "react";
import { Button, Typography } from "@mui/material";
import { BrowserBot } from "../interfaces/bot";
import { useSelector } from "react-redux";

export const BotOperation = () => {
  const [bot, setBot] = useState(null);
  const [started, setStarted] = useState(false);
  const commands = useSelector((state) => state.bot.commands);
  const token = useSelector((state) => state.bot.token);

  useEffect(() => {
    setBot(new BrowserBot(token));
  }, [token]);

  useEffect(() => {
    if (bot === null) return;
    for (const { command, response } of commands) {
      bot.addCommand(command, () => response);
    }
  }, [bot, commands]);

  return (
    <Fragment>
      <Typography variant="body1">
        Bot is {started ? "started" : "stopped"}
      </Typography>
      <Button
        variant="contained"
        color="success"
        disabled={started}
        onClick={() => {
          bot.start();
          setStarted(true);
        }}
      >
        Start
      </Button>
      <Button
        variant="contained"
        color="error"
        disabled={!started}
        onClick={() => {
          bot.stop();
          setStarted(false);
        }}
      >
        Stop
      </Button>
    </Fragment>
  );
};
