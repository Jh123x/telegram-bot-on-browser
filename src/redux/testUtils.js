import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { botSlice, defaultBotState } from "./botSlice";

export const generateDefaultState = () => {
  return {
    bot: defaultBotState,
  };
};

export const setupStore = (preloadedState) => {
  return configureStore({
    reducer: {
      bot: botSlice.reducer,
    },
    preloadedState,
  });
};

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { bot: botSlice.reducer },
      preloadedState: preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
