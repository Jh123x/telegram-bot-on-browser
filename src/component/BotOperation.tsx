import { useState, useEffect } from "react";
import { Fragment } from "react";
import { Button, Typography } from "@mui/material";
import { BrowserBot } from "../interfaces/bot.ts";
import { useDispatch, useSelector } from "react-redux";
import React from "react";
import { BotWithConfig, Command } from "../redux/types";
import { addResponse } from "../redux/botSlice.ts";

export const BotOperation = () => {
  const dispatch = useDispatch()
  const [bot, setBot] = useState<BrowserBot>();
  const [started, setStarted] = useState(false);
  const commands = useSelector<BotWithConfig, Command[]>((state) => state.bot.commands);
  const token = useSelector<BotWithConfig, string>((state) => state.bot.token);

  useEffect(() => setBot(new BrowserBot(token)), [token]);

  useEffect(() => {
    if (bot === undefined) return;
    commands.forEach(({ command, response }) => bot!.addCommand(command, () => response))
  }, [bot, commands]);

  return (
    <Fragment>
      <Typography variant="body1">
        Bot {started ? "started" : "stopped"}
      </Typography>
      <Button
        variant="contained"
        color="success"
        disabled={started}
        onClick={() => {
          bot!.start((date, user, id, msg) => {
            dispatch(addResponse({
              FromUser: user,
              UserID: id,
              Message: msg,
              Date: date,
            }))
          });
          setStarted(true);
          console.log("Bot started!")
        }}
      >
        Start
      </Button>
      <Button
        variant="contained"
        color="error"
        disabled={!started}
        onClick={() => {
          bot!.stop();
          setStarted(false);
          console.log("Bot stopped!")
        }}
      >
        Stop
      </Button>
    </Fragment >
  );
};
