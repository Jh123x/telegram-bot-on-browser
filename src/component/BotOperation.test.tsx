import { test, expect } from "@jest/globals";
import React from "react";
import { BotOperation } from "./BotOperation.tsx";
import { generateDefaultState, renderWithProviders, setupStore } from "../redux/testUtils.tsx";

test("renders bot operation interface correctly", () => {
    const store = setupStore(generateDefaultState());
    const component = renderWithProviders(<BotOperation />, { store: store });
    expect(component).toMatchSnapshot();
});
