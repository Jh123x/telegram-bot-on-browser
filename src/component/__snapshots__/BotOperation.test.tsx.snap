// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders bot operation interface correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        Bot 
        stopped
      </p>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess css-1elmdwi-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Start
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorError Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorError css-1f7qp0x-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        Stop
      </button>
    </div>
  </body>,
  "container": <div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
    >
      Bot 
      stopped
    </p>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess css-1elmdwi-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Start
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorError Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorError css-1f7qp0x-MuiButtonBase-root-MuiButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      Stop
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
